### ISSUE TITLE
Disable some "defaults", like local.{ext} loading, or specify which files you want to load.

### ISSUE THREAD

**Issue by marlon-sousa** [node-config/node-config#756]
**Is your feature request related to a problem? Please describe.**

I am using this libary as base for my own library.

My library restricts some what the available options of this one, as to enforce some standards.

My problem is the following: I would like to disable loading of all kinds of files which are loaded if present, as for example the local.{ext} ones.

Hosts and deployments files are also not desirable in my use case.

**Describe the solution you'd like**

I would create a second parameter for util.loadFileConfigs, an "options" parameter. This might be an object , default {}, which could allow an ignoreFiles = [] key.

This key would accept a list of pre determined values, to specify files to be ignored while loading:

* "DEFAULT": if present, ignores default.EXT and default-{instance}.EXT
* "DEPLOIMENT": if present, ignores deployment.EXT and deployment-{instance}.EXT
* "SHORT_HOST": if present, ignores {short_hostname}.EXT, {short_hostname}-{instance}.EXT, {short_hostname}-{deployment}.EXT 
and {short_hostname}-{deployment}-{instance}.EXT
* "FULL_HOST": if present, ignores {full_hostname}.EXT {full_hostname}-{instance}.EXT, {full_hostname}-{deployment}.EXT, {full_hostname}-{deployment}-{instance}.EXT
* "HOST": if present, ignores what is ignored by "SHORT_HOST" and "FULL_HOST"
* "LOCAL": if present, ignores local.EXT, local-{instance}.EXT, local-{deployment}.EXT and local-{deployment}-{instance}.EXT

**Describe alternatives you've considered**

As it is, there is nothing I can do about it, at least not that I can think of. I am opemning this issue because an user of my library reported that they have a local.{ext} file which was accidentally loaded when they didn't want it.

**Please tell us about your environment:**

 -  node-config version: 3.3.9
 -  node-version: from 12 to 18
 
**Other information**

I am whiling to try a pr but only if such feature would be accepted. My understanding is that because we would create an empty by default options object this wouldn't break compatibility, yet it would add functionality for those who need it.

**Comment by norbertroamsys** [comment]
We have more or less the same or an equivalent issue:
On one of our servers the hostname is the same as one of the NODE_ENV names.
As a consequence the "unwanted" hostname config file is loaded instead of the NODE_ENV config file (as documented in [File load order](https://github.com/node-config/node-config/wiki/Configuration-Files)).

An option (as descripted by @marlon-sousa) to e.g. exclude the hostname files from being loaded would help!

**Comment by markstos** [comment]
This came up before in https://github.com/node-config/node-config/issues/337

There I'm not a far of declaring what you aren't doing-- in this case, declaring what you are ignoring. It would be clear if there was a way to declare what you wanted to load, rather than what you don't want to load.

Why not just override `util.loadFileConfigs` in you private class so that it loads exactly what you want?

**Comment by norbertroamsys** [comment]
Yes - a way or strategy to define just what we wanted to be loaded would be the best. Options for ignoring are mostly a workaround for sure!

What do you mean with ...
@Why not just override util.loadFileConfigs in you private class so that it loads exactly what you want?
It also sounds like a dirty workaround which may make problems in further node-config releases.
If no: Can you provide a code snippet or an example for a proper usage?

Thanks for helping!

**Comment by marlon-sousa** [comment]
This also sounds a dangerous workaround to me.

So we are ditching a workaround for another workaround ... it seems that may be the correct way of doing that would be providing a kind of extension point like a filter that the library would call offering all possible options and the library user would return true or false for loading or not that item.

This would place a good boundary between what the library is doing and what can be configured by its users.

**Comment by markstos** [comment]
@marlon-sousa: Maybe you  are suggesting a way to declare which file or names extensions or would be processed, that would better than declaring what extensions or file names to ignore.

The larger issue is that the module needs maintenance help. There's a shortage of people triaging issues and dealing with PRs.

**Comment by jdmarshall** [comment]
I'd like to comment on the workarounds without judging the feature request on its merits.

I've found a lot of value in adding my own tests to make stronger assertions about the contents of the config files prior to running any integration tests or poorly mocked unit tests.  It's too late an confusing for your coworkers to encounter a problem caused by rotten configuration during a build, during E2E tests, or god forbid while running the application.

If you have additional limitations you want for your projects that go above and beyond what node-config allows, you might be better off failing the build if for instance people start checking in config files in a format other than the approved ones. There's a PR I filed that never landed that was intended to fix a key collision problem if you have keys named things like "get". It didn't land, so I fixed a couple of attributes that were running into it and then made a test that checked for keys that run into the bug.

Local files are good for troubleshooting and also a reasonable spot to put build metadata into (eg for a docker build), but if it's more than a few lines long I tend to agree it's questionable to have it.
